### PR TITLE
planner: support hint for IndexHashJoin and IndexMergeJoin

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
~1. fix issue https://github.com/pingcap/tidb/issues/12309~
2. support hint for IndexHashJoin and IndexMergeJoin

### Check List 

Tests 

 - Integration test

Code changes

N/A

Side effects

N/A

Related changes

N/A

Release note

N/A

### CODE REVIEW

File: executor/index_lookup_join_test.go [hunk 1]
```diff
@@ -79,6 +79,41 @@ func (s *testSuite1) TestBatchIndexJoinUnionScan(c *C) {
 	tk.MustExec("rollback")
 }
 
+func (s *testSuite1) TestIndexJoinHint(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec(`drop table if exists t1, t2;`)
+	tk.MustExec(`create table t1(a bigint, b bigint, index idx_a(a), index idx_b(b));`)
+	tk.MustExec(`create table t2(a bigint, b bigint, index idx_a(a), index idx_b(b));`)
+	tk.MustQuery("explain select /*+ INL_JOIN(t1) */ * from t1 join t2 on t1.a = t2.a;").Check(testkit.Rows(
```
Comment (lamxTyler) [inline executor/index_lookup_join_test.go]: Put them in planner test and use test file so others can update them easily?

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -1177,36 +1204,77 @@ func (p *LogicalJoin) tryToGetIndexJoin(prop *property.PhysicalProperty) (indexJ
...
+		allLeftOuterJoins = p.getIndexJoinByOuterIdx(prop, 0)
+		forcedLeftOuterJoins = make([]PhysicalPlan, 0, len(allLeftOuterJoins))
+		for _, j := range allLeftOuterJoins {
+			switch j.(type) {
+			case *PhysicalIndexJoin:
+				if hasINLJHint {
+					forcedLeftOuterJoins = append(forcedLeftOuterJoins, j)
+				}
+			case *PhysicalIndexHashJoin:
+				if hasINLHJHint {
```
Comment (lamxTyler) [inline planner/core/exhaust_physical_plans.go]: If we have inlj hint, it is also forced?

File: cmd/explaintest/r/tpch.result [hunk 3]
```diff
@@ -769,14 +769,16 @@ l_shipmode;
...
-      └─IndexMergeJoin_22	10023369.01	root	inner join, inner:TableReader_20, outer key:Column#10, inner key:Column#1
-        ├─TableReader_36	10023369.01	root	data:Selection_35
-        │ └─Selection_35	10023369.01	cop[tikv]	ge(Column#22, 1997-01-01 00:00:00.000000), in(Column#24, "RAIL", "FOB"), lt(Column#20, Column#21), lt(Column#21, Column#22), lt(Column#22, 1998-01-01)
-        │   └─TableScan_34	300005811.00	cop[tikv]	table:lineitem, range:[-inf,+inf], keep order:false
-        └─TableReader_20	1.00	root	data:TableScan_19
-          └─TableScan_19	1.00	cop[tikv]	table:orders, range: decided by [Column#10], keep order:true
+  └─HashAgg_14	1.00	root	group by:Column#43, funcs:sum(Column#40), sum(Column#41), firstrow(Column#42)
+    └─Projection_54	10023369.01	root	cast(case(or(eq(Column#6, "1-URGENT"), eq(Column#6, "2-HIGH")), 1, 0)), cast(case(and(ne(Column#6, "1-URGENT"), ne(Column#6, "2-HIGH")), 1, 0)), Column#24, Column#24
+      └─IndexMergeJoin_36	10023369.01	root	inner join, inner:Projection_34, outer key:Column#1, inner key:Column#10
```
Comment (lzmhhh123) [inline cmd/explaintest/r/tpch.result]: The inner and outer children have been swapped. Is it better?

File: planner/core/exhaust_physical_plans.go [hunk 4]
```diff
@@ -1152,18 +1152,45 @@ func (ijHelper *indexJoinBuildHelper) buildTemplateRange(matchedKeyCnt int, eqAn
...
+	inlmjLeftOuter := (p.preferJoinType & preferRightAsINLMJInner) > 0
+	hasINLMJHint := inlmjLeftOuter || inlmjRightOuter
+
+	forceLeftOuter := inljLeftOuter || inlhjLeftOuter || inlmjLeftOuter
+	forceRightOuter := inljRightOuter || inlhjRightOuter || inlmjRightOuter
 
 	defer func() {
-		if !forced && hasIndexJoinHint {
+		// refine error message
+		if !forced && (hasINLJHint || hasINLHJHint || hasINLMJHint) {
```
Comment (lzmhhh123) [inline planner/core/exhaust_physical_plans.go]: If both `hasINLJHint` and `hasINLHJHint` are `true`. Do we only generate one `errMsg` here?

File: planner/core/logical_plan_builder.go [hunk 5]
```diff
@@ -55,8 +55,12 @@ const (
 	HintSMJ = "sm_join"
 	// TiDBIndexNestedLoopJoin is hint enforce index nested loop join.
 	TiDBIndexNestedLoopJoin = "tidb_inlj"
-	// HintINLJ is hint enforce index nested loop join.
+	// HintINLJ is hint enforce index nested loop hash/sort merge join.
```
Comment (lzmhhh123) [inline planner/core/logical_plan_builder.go]: Incorrect comments?

File: planner/core/exhaust_physical_plans.go (hunk 2 above)
Comment (XuHuaiyu) [reply planner/core/exhaust_physical_plans.go]: Yes

File: cmd/explaintest/r/explain_complex.result [hunk 6]
```diff
@@ -152,14 +152,15 @@ explain select dt.id as id, dt.aid as aid, dt.pt as pt, dt.dic as dic, dt.cm as
...
 └─Limit_13	0.00	root	offset:0, count:2000
-  └─IndexJoin_19	0.00	root	inner join, inner:IndexLookUp_18, outer key:Column#2, Column#5, inner key:Column#21, Column#23
-    ├─TableReader_43	0.00	root	data:Selection_42
-    │ └─Selection_42	0.00	cop[tikv]	eq(Column#18, 0), eq(Column#4, "ios"), gt(Column#9, 1478185592), not(isnull(Column#5))
-    │   └─TableScan_41	10000.00	cop[tikv]	table:dt, range:[0,+inf], keep order:false, stats:pseudo
-    └─IndexLookUp_18	0.00	root	
-      ├─IndexScan_15	1.25	cop[tikv]	table:rr, index:aid, dic, range: decided by [eq(Column#21, Column#2) eq(Column#23, Column#5)], keep order:false, stats:pseudo
-      └─Selection_17	0.00	cop[tikv]	eq(Column#22, "ios"), gt(Column#26, 1478185592)
-        └─TableScan_16	1.25	cop[tikv]	table:rr, keep order:false, stats:pseudo
+  └─IndexMergeJoin_41	0.00	root	inner join, inner:IndexLookUp_39, outer key:Column#21, Column#23, inner key:Column#2, Column#5
```
[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply cmd/explaintest/r/explain_complex.result]: fixed

File: cmd/explaintest/r/tpch.result (hunk 3 above)
Comment (XuHuaiyu) [reply cmd/explaintest/r/tpch.result]: fixed

File: planner/core/exhaust_physical_plans.go (hunk 4 above)
Comment (XuHuaiyu) [reply planner/core/exhaust_physical_plans.go]: https://github.com/pingcap/tidb/pull/13238/files#diff-b9e75a99170ea35a67074d5d3d3305adR411

File: planner/core/exhaust_physical_plans.go (hunk 2 above)
Comment (lamxTyler) [reply planner/core/exhaust_physical_plans.go]: But it is not currently?

File: planner/core/physical_plan_test.go [hunk 7]
```diff
@@ -683,6 +685,7 @@ func (s *testPlanSuite) TestHintScope(c *C) {
 		stmt, err := s.ParseOneStmt(test, "", "")
 		c.Assert(err, IsNil, comment)
 
+		logrus.Warning("test:------", test)
```
Comment (lamxTyler) [inline planner/core/physical_plan_test.go]: Remove it.

File: planner/core/exhaust_physical_plans.go [hunk 8]
```diff
@@ -1182,36 +1209,93 @@ func (p *LogicalJoin) tryToGetIndexJoin(prop *property.PhysicalProperty) (indexJ
...
+					forcedLeftOuterJoins = append(forcedLeftOuterJoins, j)
+				}
+			}
 		}
-
-		rightJoins := p.getIndexJoinByOuterIdx(prop, 1)
-		if rightJoins != nil && (rightOuter && !leftOuter || rhsCardinality < lhsCardinality) {
-			return rightJoins, rightOuter
+		switch {
+		case p.JoinType == InnerJoin && p.Children()[0].statsInfo().Count() < p.Children()[1].statsInfo().Count():
```
Comment (lzmhhh123) [inline planner/core/exhaust_physical_plans.go]: Well, I don't think it's always a better choice for the situation to force left outers. What about remove the second condition?

Comment (lzmhhh123) [reply planner/core/exhaust_physical_plans.go]: Then we can get both sides inner join.

Comment (XuHuaiyu) [reply planner/core/exhaust_physical_plans.go]: It'll affect many plans, using an individual PR is better. [This issue](https://github.com/pingcap/tidb/issues/12309) traces it.

**Pull request merged**